common/TextDifferencer.java

```diff
@@ -19,30 +19,108 @@
-            computeLCSMatrix(first.toCharArray(), second.toCharArray()),
-            first.toCharArray(),
-            second.toCharArray())
-        .stream()
-        .map((builder) -> builder.build())
+    int headerLength = getHeaderMatchingCharactersLength(first.toCharArray(), second.toCharArray());
+    int footerLength =
+        getFooterMatchingCharactersLength(
+            first.substring(headerLength).toCharArray(),
+            second.substring(headerLength).toCharArray());
+    char[] firstBody = first.substring(headerLength, first.length() - footerLength).toCharArray();
+    char[] secondBody =
+        second.substring(headerLength, second.length() - footerLength).toCharArray();
+    return mergeDifferences(
+        getMatchingCharDifferences(first.toCharArray(), 0, headerLength),
+        addOffestTo(getNonMatchingCharDifferences(firstBody, secondBody), headerLength),
+        getMatchingCharDifferences(
+            first.toCharArray(), first.length() - footerLength, footerLength));
+  }
+
+  /** Merge all the differences into a single {@link List}. */
+  private static List<CharDifference> mergeDifferences(
+      Stream<CharDifference.Builder> headerDifferences,
+      Stream<CharDifference.Builder> bodyDifference,
+      Stream<CharDifference.Builder> footerDifferences) {
+    return Stream.of(headerDifferences, bodyDifference, footerDifferences)
+        .flatMap(Function.identity())
+        .map(builder -> builder.build())
         .collect(Collectors.toList());
   }
 
+  /** Count the number of equal characters from the beginning of the given two strings. */
+  private static int getHeaderMatchingCharactersLength(final char[] first, final char[] second) {
+    int count = 0;
+    for (; count < first.length && count < second.length; count++) {
+      if (first[count] != second[count]) {
+        return count;
+      }
+    }
+    return count;
+  }
+
+  /** Count the number of equal characters from the end of the given two strings. */
+  private static int getFooterMatchingCharactersLength(final char[] first, final char[] second) {
+    int count = 0;
+    for (; count < first.length && count < second.length; count++) {
+      if (first[first.length - count - 1] != second[second.length - count - 1]) {
+        return count;
+      }
+    }
+    return count;
+  }
+
+  /**
+   * Correct the offset to the given {@link CharDifference.Builder}.
+   *
+   * @param stream A stream of char difference builder.
+   * @param offset The offset required for correction.
+   * @return A stream with the corrected offset.
+   */
+  private static Stream<CharDifference.Builder> addOffestTo(
+      final Stream<CharDifference.Builder> stream, final int offset) {
+    return stream.map(diff -> diff.setIndex(offset + diff.getIndex()));
+  }
+
+  /** Return a stream of the non equal char differences of the given two strings. */
+  private static Stream<CharDifference.Builder> getNonMatchingCharDifferences(
```

Comment:
I think that if you can make `getDifferencesFromLCSMatrix` work with `int begin, int length`, then you won't need `getNonMatchingCharDifferences` and `addOffestTo`.

Then, if you return a list from `getMatchingCharDifferences`, you can directly add to a list using `addAll` and won't need `mergeDifferences`.